### PR TITLE
Locate `kotlinc` filepath via IntelliJ IDEA's bundled Kotlin plugin

### LINKED ISSUE #311: Compilation for Kotlin does not work if `kotlinc` is absent in `PATH`
**Describe the bug**
I cannot generate unit tests with LLM for Kotlin if I do not set `kotlinc` in `PATH`. The compiler fails with:
```log
'bash: kotlinc: command not found'
```
This is because, unlike for Java, for Kotlin we do not search for compiler installation but rely on its presence in `PATH`:
```kotlin
// in KotlinTestCompiler.kt

val errorMsg = CommandLineRunner.run(
            arrayListOf(
                "kotlinc", // <- HERE! but in `JavaTestCompiler` we search for the javac
                "-cp",
                classPaths,
                path,
            ),
        )
```
The plugin does not notify that compiler was not found and misleadingly treat it as a compilation error:
Screenshot 2024-07-29 at 12 48 07

**To Reproduce**
1. Remove the `kotlinc` from your `PATH`.
2. Try to generate tests with LLM.
3. See the compilation error being `"kotlinc not found"`.

**Expected behavior**
TestSpark should search for `kotlinc` as it does for Java.

### PR DESCRIPTION
**Pull request by Vladislav0Art**
# Description of changes made
Retrieves kotlin compiler (`kotlinc`) via `KotlinPluginLayout` component of `org.jetbrains.kotlin` bundled plugin.

# Why is merge request needed
Closes #311 

*Please write if you have anything to add*

# What is missing?
* [x] Test the solution on Windows (test whether Windows has ~~`kotlinc.exe`~~ (it has `kotlinc`) as a compiler):
     1. In the case of Windows, `kotlinc` requires `java` to be executable (e.g., by being present in `PATH`); otherwise, `kotlinc` will complain about missing `java`.
     2. In other words, Kotlin compilation works on Windows but with a nuance. My Windows machine has an old version of IntelliJ (see comment in a file). Maybe newer versions solve this issue. 

* [ ] Once the PR is reviewed, I will publish a new version of `core` module.

## Self-check
- [x] I have checked that I am merging into the correct branch

### CODE REVIEW

**Review by pderakhshanfar — CHANGES_REQUESTED** [review]
1. RuntimeExceptions in `KotlinTestCompiler`, `JavaTestCompiler`, and `TestCompilerFactory` shoudl be handled properly.
2. Since we are using path to `kotlinc`, the if condition for checking the error message [here](https://github.com/JetBrains-Research/TestSpark/blob/5975b97b5238ac2755c76d53405b09deef63371f/core/src/main/kotlin/org/jetbrains/research/testspark/core/test/kotlin/KotlinTestCompiler.kt#L67) might not work as expected.

**Review by arksap2002 — CHANGES_REQUESTED** [review]
Remaining message: https://github.com/JetBrains-Research/TestSpark/pull/376#issuecomment-2400093853

File: src/main/kotlin/org/jetbrains/research/testspark/tools/factories/TestCompilerFactory.kt [hunk 1]
```diff
@@ -16,17 +17,38 @@ object TestCompilerFactory {
...
+        project: Project,
+        libraryPaths: List<String>,
+        junitLibraryPaths: List<String>,
+        javaHomeDirectory: String? = null,
+    ): JavaTestCompiler {
+        val javaSDKHomePath = javaHomeDirectory
+            ?: ProjectRootManager.getInstance(project).projectSdk?.homeDirectory?.path
+
+        if (javaSDKHomePath == null) {
+            throw RuntimeException("Java SDK not configured for the project.")
```
[the remark this answers is not in the document]
Comment (Vladislav0Art) [reply src/main/kotlin/org/jetbrains/research/testspark/tools/factories/TestCompilerFactory.kt]: Now, throwing a custom exception that is handled in `Pipeline` and `Llm`.

**Comment by Vladislav0Art** [comment]
> @Vladislav0Art where do you catch new `RuntimeException`?

[Solved](https://github.com/JetBrains-Research/TestSpark/pull/376#discussion_r1794233202).

**Comment by Vladislav0Art** [comment]
> 2\. Since we are using path to `kotlinc`, the if condition for checking the error message [here](https://github.com/JetBrains-Research/TestSpark/blob/5975b97b5238ac2755c76d53405b09deef63371f/core/src/main/kotlin/org/jetbrains/research/testspark/core/test/kotlin/KotlinTestCompiler.kt#L67) might not work as expected

@pderakhshanfar 

This error message is outputted by a CLI, not kotlinc. I removed it because the custom exception thrown above in the init block does not allow `kotlinc` to be absent (because we literally search for an executable with this name; if it was not found in the init block, we'd throw an exception there).

**Review by arksap2002 — APPROVED** [review]

**Review by pderakhshanfar — APPROVED** [review]

**Pull request merged**